### PR TITLE
Firmware modified to support CODAL

### PR DESCRIPTION
**Pull request by martinwork**
microbit-v2-samples is included as a submodule to build for V2. 

After cloning the repo, to build V2 firmware microbit-firmata/microbit-v2-samples/MICROBIT.hex:
cd microbit-firmata
git submodule update --init
cd firmware
python buildv2.py

The desktop client expects some event source numbers that are different in V1 and V2. The CODAL firmware translates the V2 source numbers to the V1 ones in eventToExternal(). This is OK because only a few events are supported, but could become very awkward if more events are supported.

The code supports macro FIRMATA_USE_UBIT to test V1 and V2 compiling either uBit, or just the required components (the default), but with neither option starting the scheduler. The streaming test speeds are affected very slightly (0.1% ?) by compiling uBit.

### CODE REVIEW

**Comment by jaustin** [comment]
@martinwork it sounds like it might be a better step then to update the numbers used by the desktop software for the events and align DAL and CODAL? Is this something @finneyj did for MakeCode too? Or is there some translation there?

**Comment by finneyj** [comment]
I thought we had aligned most of them already - do we know which ones are different?

**Comment by martinwork** [comment]
@jaustin @finneyj 

The external clients are expecting the DAL numbers:
https://github.com/microbit-foundation/microbit-firmata/blob/master/client/MBFirmataClient.js#L322

I was assuming there were clients we didn't have control over, so it would be CODAL changing to match DAL.

What DAL update aligned the IDs? Is it a difference between DAL master and the branch used for MakeCode?

These are the IDs the V2 firmware is translating from CODAL to the one's expected by the client:
https://github.com/microbit-foundation/microbit-firmata/pull/12/files#diff-3586b71736669d8789c831e156f52d0f5e6ba35a8c8b1371e402313746f2ce6fR141

**Comment by finneyj** [comment]
I think it's this one:
https://github.com/lancaster-university/microbit-dal/commit/76538ef6cb392f94d4402bdf26f5a0e581246520

Can't remember why we did this precisely - but was at the request of makeCode team. Probably because on v2 we use pxt-common-packages much more...

**Comment by martinwork** [comment]
That is in the js-event-semantics branch, right? 

@jaustin Should I point the firmata V1 build at v2.2.0-rc6, and update the constants in MBFirmataClient.js?

**Comment by finneyj** [comment]
Yes, MakeCode for micro:bit v1 is running off that branch AFAIK.

**Comment by jaustin** [comment]
@martinwork yes, I think so. We'll need to make a new release that's a breaking change with these IDs but I think the only sane way to do it is to align V1 and V2 on the same IDs, and I think the only way we could do that sanely for pxt was by bringing V1 up to V2 so that's what had to happen. @finneyj I guess this will affect other users of CODAL - do we need to think about a compat system? Is there a 1:1 mapping for V1:V2 IDs that we can use? What's the overlap between the V2 range and V1 range? Could V2 additionally support all V1 IDs mapped to the new namespace?

**Comment by martinwork** [comment]
@jaustin I don't know what other clients are out there, or if they are all based on MBFirmataClient.js. If a client is changed to use new numbers, old firmware will stop working.

We can avoid any need for changes by simply deciding / defining that the message MB_REPORT_EVENT will only ever be used for the current 7 events with the V1 numbers. If required in the future, a new message could support other raw events.

Perhaps all we need is a comment to that effect in the firmware code.

**Comment by jaustin** [comment]
@epeach could you let us know whether it'd be an issue at your end to renumber the events in line with the newer numbers? Although we *could* maintain the older ones, @martinwork as I think there are few users of this so far, I'd be inclined to introduce things that look like 'legacy' at this point and perhaps do a protocol rev that allows anyone who is using it to determine that the numbers have changed?

**Comment by martinwork** [comment]
@jaustin I didn't realise that the number of simple messages is quite limited. 
https://github.com/firmata/protocol/blob/master/protocol.md
https://github.com/microbit-foundation/microbit-firmata/blob/master/firmware/source/mbFirmata.h#L56

We can change the firmware version:
https://github.com/microbit-foundation/microbit-firmata/blob/master/firmware/source/mbFirmata.cpp#L143

MBFirmataClient.js doesn't use the firmware version at the moment apart from storing it in a string.
https://github.com/microbit-foundation/microbit-firmata/blob/master/client/MBFirmataClient.js#L290

So MBFirmataClient.js and any other clients will have to store the firmware version numbers and expect different IDs starting with v1.1.

**Comment by martinwork** [comment]
Added new commit https://github.com/microbit-foundation/microbit-firmata/pull/12/commits/771f2d855b3665b1f5c71a235e722661f5e71c47

- use DAL v2.2.0-rc6
- move varying event ID code to client
- bug fixes in the client and tests
- V2 build instructions in README
- Universal microbit-firmata-v1.1.hex in precompiled folder

**Comment by jaustin** [comment]
@epeach could you please try this image?  (the raw download from https://github.com/microbit-foundation/microbit-firmata/blob/771f2d855b3665b1f5c71a235e722661f5e71c47/precompiled/microbit-firmata-v1.1.hex should work)

**Comment by martinwork** [comment]
@jaustin @epeach

Client code must be updated to use the event numbers that vary between v1.0 and v1.1.
This PR updates MBFirmataClient.js for this.

Any event handler registered with MBFirmataClient.js should use the variables set here:
https://github.com/microbit-foundation/microbit-firmata/blob/771f2d855b3665b1f5c71a235e722661f5e71c47/client/MBFirmataClient.js#L116

Like this:
https://github.com/microbit-foundation/microbit-firmata/blob/771f2d855b3665b1f5c71a235e722661f5e71c47/client/mbTests.js#L405

**Comment by epeach** [comment]
Hi there - not a problem on our end to renumber the events. I tried the v1.1.hex you linked and after a quick look at a sample project, it appears to be working as expected. Let me know if there's any other questions!

**Comment by martinwork** [comment]
Thanks @epeach. I have fixed the ID in the "fake" animation complete event the firmware sends when processing a display request while the display is disabled, and updated the precompiled hex.

File: client/MBFirmataClient.js [hunk 1]
```diff
@@ -181,6 +201,8 @@ class MicrobitFirmataClient {
 		var id = usbSerialNumber.slice(0, 4);
 		if ('9900' == id) return '1.3';
 		if ('9901' == id) return '1.5';
+		if ('9903' == id) return '2.0';
+		if ('9904' == id) return '2.0';
```
Comment (microbit-carlos) [inline client/MBFirmataClient.js:205]: @JohnVidler We'll need to add board IDs 9905 and 9906 here as well for V2.2.

**Pull request merged**